src/factorization.jl

```diff
@@ -1242,37 +1318,44 @@ function FastQRFactorization()
 end
 
 @static if VERSION < v"1.7beta"
-    function init_cacheval(alg::FastQRFactorization{Val{false}}, A, b, u, Pl, Pr,
+    function init_cacheval(alg::FastQRFactorization{Val{false}}, A::AbstractMatrix, b, u, Pl, Pr,
         maxiters::Int, abstol, reltol, verbose::Bool,
         assumptions::OperatorAssumptions)
         ws = QRWYWs(A; blocksize = alg.blocksize)
         return WorkspaceAndFactors(ws,
-            ArrayInterface.qr_instance(convert(AbstractMatrix, A)))
+            ArrayInterface.qr_instance(A))
     end
 
-    function init_cacheval(::FastQRFactorization{Val{true}}, A, b, u, Pl, Pr,
+    function init_cacheval(::FastQRFactorization{Val{true}}, A::AbstractMatrix, b, u, Pl, Pr,
```

Comment:
what about the operator dispatches?

Reply:
That's only for matrix operator. What about other abstract scimloperators?

Reply:
There is none for that,
but note that this is currently not working anyway:

```julia
using LinearSolve
using SciMLOperators

N = 4
f = (u, p, t) -> u .* u

M = MatrixOperator(rand(N, N))
D = DiagonalOperator(rand(N))
F = FunctionOperator(f, zeros(N), zeros(N))

b = rand(4)

solve(LinearProblem(rand(N, N), b),FastQRFactorization()) # works
solve(LinearProblem(M, b),FastQRFactorization()) # fails
solve(LinearProblem(D, b),FastQRFactorization()) # fails
solve(LinearProblem(F, b),FastQRFactorization()) # fails
```

julia> solve(LinearProblem(M, b),FastQRFactorization())
ERROR: MethodError: init_cacheval(::FastQRFactorization{LinearAlgebra.NoPivot}, ::MatrixOperator{Float64, Matrix{Float64}, SciMLOperators.FilterKwargs{typeof(SciMLOperators.DEFAULT_UPDATE_FUNC), Tuple{}}, SciMLOperators.FilterKwargs{typeof(SciMLOperators.DEFAULT_UPDATE_FUNC), Tuple{}}}, ::Vector{Float64}, ::Vector{Float64}, ::IdentityOperator, ::IdentityOperator, ::Int64, ::Float64, ::Float64, ::Bool, ::OperatorAssumptions{Bool}) is ambiguous.

Candidates:
  init_cacheval(alg::FastQRFactorization, A::MatrixOperator, b, u, Pl, Pr, maxiters::Int64, abstol, reltol, verbose::Bool, assumptions::OperatorAssumptions)
    @ LinearSolve C:\Users\arno\.julia\packages\LinearSolve\8EgR1\src\factorization.jl:1180
  init_cacheval(alg::FastQRFactorization{LinearAlgebra.NoPivot}, A, b, u, Pl, Pr, maxiters::Int64, abstol, reltol, verbose::Bool, assumptions::OperatorAssumptions)
    @ LinearSolve C:\Users\arno\.julia\packages\LinearSolve\8EgR1\src\factorization.jl:1066

Possible fix, define
  init_cacheval(::FastQRFactorization{LinearAlgebra.NoPivot}, ::MatrixOperator, ::Any, ::Any, ::Any, ::Any, ::Int64, ::Any, ::Any, ::Bool, ::OperatorAssumptions)

Reply:
For it to ever have worked for FunctionOperators it needed to have this defined
```jl
WorkspaceAndFactors(ws, ArrayInterface.qr_instance(A))
```

Anyway, the solution for DiagonalOperators seems straightforward to me. Just add the same block of code as for MatrixOperators. But how do you want to handle FunctionOperators?